### ISSUE TITLE
javascript Promise

### ISSUE THREAD

**Issue by cercier** [cercier/some_notes_in_learning_python#128]
### javascript language

- single-threaded
- event-driven
- non-blocking I/O model
- Web/Node API (setTimeout, DOM, XmlHttpRequest...)

The asynchronous behaviour is not part of the JavaScript language itself, but built on top of the core JavaScript language in the browser (or the programming environment) and accessed through the Web/Node API. Asynchronous tasks are run in the background either by the browser on the web, or by the C++ layer of NodeJS.

### Call Stack

- LIFO queue
- If that function calls another function within it then that function gets called to the top of the stack.
- execute one task, or piece of code at a time

The ***Event Loop*** continuously checks the **_Call Stack_** to see if there’s any function that needs to run.

Example
```js
const bar = () => console.log('bar');
const baz = () => console.log('baz');
const foo = () => {
  console.log('foo');
  bar();
  baz();
}
foo();
```
prints
```
foo
bar
baz
```

![call stack](https://flaviocopes.com/javascript-event-loop/call-stack-first-example.png)
### Event Loop

- a constantly running **process** that checks if the call stack is empty

The increment that the event loop moves in is called a ‘***tick***’, and every time it ‘ticks’ it checks if the call stack is empty, if so , it adds the top function in the event queue to the call stack and executes it. Once it is finished processing this function it starts ticking again.

### Message Queue(Event Queue)

JavaScript runtimes contain a ***Message Queue*** which stores **a list of messages** to be processed and their associated **callback functions**.

Example
```js
const bar = () => console.log('bar');
const baz = () => console.log('baz');
const foo = () => {
  console.log('foo');
  
  //setTimeout is pushed into the call stack and execution starts. Once the handover to browser
  // is complete, it's pop out from the stack. Once the browser receives it, the callback will 
  //be added to the message queue immediately as delay is 0.
  setTimeout(bar, 0) ;
  baz();
}
foo();
```
prints
```
foo
baz
bar
```
When ***setTimeout()*** is called, the Browser or Node.js start the timer. Once the timer expires, in this case immediately as we put 0 as the timeout, the callback function is put in the ***Message Queue***.

>***setTimeout***(message[,time] )
> - It's a Web/Node API.
>- The second argument(defaults to 0) indicates a minimum time and not a guaranteed time.
The ***time*** value represents the (minimum) delay after which the message will actually be pushed into the queue. If there is no other message in the ***Message Queue***, the message is processed right after the delay; however, if there are messages, the ***setTimeout*** message will have to wait for other messages to be processed.

The ***event loop*** gives priority to the call stack, and it first processes everything it finds in the ***call stack***, and once there’s nothing in there, it goes to pick up things in the event queue.

### Job Queue(microtask)

ES2015 introduced the concept of the ***Job Queue***, which is used by ***Promises***. It’s a way to execute the result of an async function as soon as possible, rather than being put at the end of the call stack.

It’s a layer on top of the Event Loop queue. ***Promises*** that resolve before the current function ends will be executed right after the current function.

Example:

```js
const bar = () => console.log('bar');
const baz = () => console.log('baz');
const foo = () => {
  console.log('foo');
  setTimeout(bar, 0);
  new Promise((resolve, reject) =>
    resolve('should be right after baz, before bar')
  ).then(resolve => console.log(resolve));
  baz();
}
foo();
```
prints
```
foo
baz
should be right after baz, before bar
bar
```
***Promise*** will add a job to the ***Job Queue***, but the ***setTimeout*** will put the function to the end of the event loop, when time expired. when a job is added to the job queue, javascript engine will take the next job on the ***Job Queue*** and only when ***Job Queue*** is empty , it moves to the next entry on the ***Event Loop***.

[recommended reading](https://jakearchibald.com/2015/tasks-microtasks-queues-and-schedules/)

### summary

[from](from%20https://stackoverflow.com/questions/49292733/how-promise-callbacks-are-scheduled-vs-settimeout-by-js-engines-or-browser-node)
![](https://i.stack.imgur.com/zCDgM.jpg)
### API

|  Static Methods |  |
|--|--|
| `Promise.resolve(promise);` |  Returns promise (only if `promise.constructor == Promise`)|
| `Promise.resolve(thenable);` | Make a new promise from the thenable. A thenable is promise-like in as far as it has a `then()` method. |
| `Promise.resolve(obj);` | Make a promise that fulfills to `obj`. in this situation. |
| `Promise.reject(obj);` | Make a promise that rejects to `obj`. For consistency and debugging (e.g. stack traces), `obj` should be an `instanceof Error`. |
| `Promise.all(array);` | Make a promise that fulfills when every item in the array fulfills, and rejects if (and when) any item rejects. Each array item is passed to `Promise.resolve`, so the array can be a mixture of promise-like objects and other objects. The fulfillment value is an array (in order) of fulfillment values. The rejection value is the first rejection value. |
| `Promise.race(array);` | Make a Promise that fulfills as soon as any item fulfills, or rejects as soon as any item rejects, whichever happens first. |

| Constructor  |  |
|--|--|
| `new Promise(function(resolve, reject) {});` |  |
| `resolve(thenable)` | Your promise will be fulfilled/rejected with the outcome of `thenable` |
| `resolve(obj)` | Your promise is fulfilled with `obj` |
| `reject(obj)` | Your promise is rejected with `obj`. For consistency and debugging (e.g., stack traces), obj should be an `instanceof Error`. Any errors thrown in the constructor callback will be implicitly passed to `reject()`. |

| Instance Methods |  |
|--|--|
| `promise.then(onFulfilled, onRejected)` | `onFulfilled` is called when/if "promise" resolves. `onRejected` is called when/if "promise" rejects. Both are optional, if either/both are omitted the next `onFulfilled`/`onRejected` in the chain is called. Both callbacks have a single parameter, the fulfillment value or rejection reason. `then()` returns a new promise equivalent to the value you return from `onFulfilled`/`onRejected` after being passed through `Promise.resolve`. If an error is thrown in the callback, the returned promise rejects with that error. |
| `promise.catch(onRejected)` |Sugar for `promise.then(undefined, onRejected)`  |

### terminology

A promise can be:

-   **fulfilled** - The action relating to the promise succeeded
-   **rejected** - The action relating to the promise failed
-   **pending** - Hasn't fulfilled or rejected yet
-   **settled** - Has fulfilled or rejected
-  **thenable** - to describe an object that is promise-like

#### XmlHttpRequest using Promise
```js
function get(url) {
  // Return a new promise.
  return new Promise(function(resolve, reject) {
    // Do the usual XHR stuff
    var req = new XMLHttpRequest();
    req.open('GET', url);

    req.onload = function() {
      // This is called even on 404 etc
      // so check the status
      if (req.status == 200) {
        // Resolve the promise with the response text
        resolve(req.response);
      }
      else {
        // Otherwise reject with the status text
        // which will hopefully be a meaningful error
        reject(Error(req.statusText));
      }
    };

    // Handle network errors
    req.onerror = function() {
      reject(Error("Network Error"));
    };

    // Make the request
    req.send();
  });
}
```
usage
```js
get('story.json').then(function(response) {
  console.log("Success!", response);
}, function(error) {
  console.error("Failed!", error);
})
```
#### chaining
```js
get('story.json').then(function(response){
    return JSON.parse(response);
}).then(function(response){
    console.log('Yey JSON!', response);
});

// shortcut since JSON.Parse takes one arg and returns a value
get('story.json').then(JSON.parse).then(function(reponse){
    console.log('Yey JSON!', reponse);
});

// getJson still returns a promise
function getJSON(url) {
  return get(url).then(JSON.parse);
}
```

#### chain `then`s to run async actions in sequence

When you return something from a `then()` callback, if you return a value, the next `then()` is called with that value. If you return something promise-like, the next `then()` waits on it, and is only called when that promise settles (succeeds/fails).

```js
getJSON('story.json').then(function(story) {
  return getJSON(story.chapterUrls[0]);
}).then(function(chapter1) {
  console.log("Got chapter 1!", chapter1);
})
```
a shortcut method to get chapters, ```story.json``` is only fetched once: 
```js
var storyPromise;
function getChapter(i){
    storyPromise= storyPromise || getJson('story.json');
    return storyPromise.then(function(story){
        return getJson(story.chapterUrls[i]);
    });
}
getChapter(0).then(function(chapter){
    console.log(chapter);
    return getChapter(1);
}).then(function(chapter){
    console.log(chapter);
});

```

####  Error handling

`catch()`:
```js
get('story.json').then(function(response){
    console.log('Success!', response);
}, function(error){
    console.log('Failed!', error);
});
```
sugar for `then(undefined, func)`: 
```js
get('story.json').then(function(response){
    console.log('Success!', response);
}).catch(function(err){
    console.log('Failed!', error);
});
```
With `then(func1, func2)`, `func1` or `func2` will be called, never both. But with `then(func1).catch(func2)`, both will be called if `func1` rejects, as they're separate steps in the chain.
```js
asyncThing1().then(function() {
  return asyncThing2();
}).then(function() {
  return asyncThing3();
}).catch(function(err) {
  return asyncRecovery1();
}).then(function() {
  return asyncThing4();
}, function(err) {
  return asyncRecovery2();
}).catch(function(err) {
  console.log("Don't worry about it");
}).then(function() {
  console.log("All done!");
})
```
![order](./thencatch.png)

### Question
1:
[from stackoverflow](https://stackoverflow.com/questions/36870467/what-is-the-order-of-execution-in-javascript-promises)
```js
Promise.resolve('A')
  .then(function(a){console.log(2, a); return 'B';})
  .then(function(a){
     Promise.resolve('C')
       .then(function(a){console.log(7, a);})
       .then(function(a){console.log(8, a);});
     console.log(3, a);
     return a;})
  .then(function(a){
     Promise.resolve('D')
       .then(function(a){console.log(9, a);})
       .then(function(a){console.log(10, a);});
     console.log(4, a);})
  .then(function(a){
     console.log(5, a);});
console.log(1);
setTimeout(function(){console.log(6)},0);
```
2:

[from jakearchibald](https://jakearchibald.com/2015/tasks-microtasks-queues-and-schedules/)
### async
`async` always returns a promise. If the code has `return <non-promise>` in it, then JavaScript automatically wraps it into a resolved promise with that value.
```javascript
async function f() {
  return 1;
}
f().then(alert); // 1
```
which is same as:
```javascript
async function f() {
  return Promise.resolve(1);
}
f().then(alert); // 1
```
### await
`await` works only inside async functions. It makes JavaScript wait until that promise settles and returns its result.
```javascript
async function f() {
  let promise = new Promise((resolve, reject) => {
    setTimeout(() => resolve("done!"), 1000)
  });
  let result = await promise; // wait till the promise resolves (*)
  alert(result); // "done!"
}
f();
```
### error handling
use `try...catch`
```javascript
async function f() {
  try {
    let response = await fetch('http://no-such-user');
    let user = await response.json();
  } catch(err) {
    // catches errors both in fetch and response.json
    alert(err);
  }
}
f();
```
or:
```javascript
async function f() {
  let response = await fetch('http://no-such-user');
  let user = await response.json();
}
// f() becomes a rejected promise
f().catch(alert); // TypeError: failed to fetch
```
### Being Careful!
execute in series:
```js
function wait(ms) {  return new Promise(r => setTimeout(r, ms));}
```
```js
async function series() {
  await wait(500); // Wait 500ms…
  await wait(500); // …then wait another 500ms.
  return "done!";
}
```
execute in parallel:
```js
async function parallel() {
  const wait1 = wait(500); // Start a 500ms timer asynchronously…
  const wait2 = wait(500); // …meaning this timer happens in parallel.
  await wait1; // Wait 500ms for the first timer…
  await wait2; // …by which time this timer has already finished.
  return "done!";
}
```

### demo
stream a response and return the final size using promise:
```js
function getResponseSize(url) {
    return fetch(url).then(response => {
        const reader = response.body.getReader();
        let bytesReceived = 0;

        return reader.read().then(function processResult(result) {
            if (result.done) {
                console.log('Fetch complete');
                return;
            }
            bytesReceived += result.value.length;
            console.log('Received bytes', bytesReceived);

            return reader.read().then(processResult);
        })
    });
}
```
using async function:
```js
async function getResponseSize(url) {
    const response = await fetch(url);
    const reader = response.body.getReader();
    let result = await reader.read();
    let bytesReceived = 0;

    while (!result.done) {
        const value = result.value;
        bytesReceived += result.value.length;
        console.log('Received bytes', bytesReceived);
        result = await reader.read();
    }
    return bytesReceived;
}
```
## async iterator and generator
an experimental feature available in Chrome Canary and node 10
```js
async function getResponseSize(url) {
  const response = await fetch(url);
  let total = 0;

  for await (const chunk of response.body) {
    total += chunk.length;
  }
  return total;
}
```
`for-of `loop will get its iterator by calling` thing[Symbol.iterator]`. Whereas the `for-await` loop will get its iterator by calling` asyncThing[Symbol.asyncIterator]` if it's defined, otherwise it will fall back to `asyncThing[Symbol.iterator]`

generator example
```js
async function* asyncRandomNumbers() {
  // This is a web service that returns a random number
  const url = 'https://www.random.org/decimal-fractions/?num=1&dec=10&col=1&format=plain&rnd=new';

  while (true) {
    const response = await fetch(url);
    const text = await response.text();
    yield Number(text);
  }
}
async function example() {
  for await (const number of asyncRandomNumbers()) {
    console.log(number);
    if (number > 0.95) break;
  }
}
```

## Thanks to
https://developers.google.com/web/fundamentals/primers/promises
https://developers.google.com/web/fundamentals/primers/async-functions
https://davidwalsh.name/promises
https://javascript.info/async-await#await
https://jakearchibald.com/2015/tasks-microtasks-queues-and-schedules/
https://jakearchibald.com/2017/async-iterators-and-generators/
https://javascript.info/async-await#await
https://stackoverflow.com/questions/49292733/how-promise-callbacks-are-scheduled-vs-settimeout-by-js-engines-or-browser-node
https://stackoverflow.com/questions/36870467/what-is-the-order-of-execution-in-javascript-promises
https://flaviocopes.com/javascript-event-loop/
https://blog.sessionstack.com/how-javascript-works-event-loop-and-the-rise-of-async-programming-5-ways-to-better-coding-with-2f077c4438b5?gi=58d2f8d9e347
